src/server/utilities.ts

```diff
@@ -232,6 +232,79 @@ namespace ts.server {
         }
     }
 
+    export function cleanExistingMap<T>(
+        existingMap: Map<T>,
+        onDeleteExistingValue: (key: string, existingValue: T) => void) {
+        if (existingMap) {
+            // Remove all
+            existingMap.forEach((existingValue, key) => {
+                existingMap.delete(key);
+                onDeleteExistingValue(key, existingValue);
+            });
+        }
+    }
+
+    export function mutateExistingMapWithNewSet<T>(
+        existingMap: Map<T>, newMap: Map<true>,
+        createNewValue: (key: string) => T,
+        onDeleteExistingValue: (key: string, existingValue: T) => void
+    ): Map<T> {
+        return mutateExistingMap(
+            existingMap, newMap,
+            // Same value if the value is set in the map
+            /*isSameValue*/(_existingValue, _valueInNewMap) => true,
+            /*createNewValue*/(key, _valueInNewMap) => createNewValue(key),
+            onDeleteExistingValue,
+            // Should never be called since we say yes to same values all the time
+            /*OnDeleteExistingMismatchValue*/(_key, _existingValue) => notImplemented()
+        );
+    }
+
+    export function mutateExistingMap<T, U>(
+        existingMap: Map<T>, newMap: Map<U>,

```

Comment:
This could make use of `ReadonlyMap` from #17161. `existingMap` is mutated, `newMap` is readonly.